File: 6.js
// LeetCode Zigzag Conversion
// Reference: https://leetcode.com/problems/zigzag-conversion/solutions/5004842/video-simple-solution/?envType=study-plan-v2&envId=top-interview-150

/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function(s, numRows) {
    if (numRows === 1 || numRows >= s.length) {
        return s;
    }

    let idx = 0, d = 1;
    const rows = new Array(numRows).fill().map(() => []);

    for (const char of s){
        rows[idx].push(char);
        if (idx === 0){
            d = 1;

        } else if (idx === numRows - 1){
            d = -1;
        }
        idx += d;
    }

    for (let i = 0; i < numRows; i++){
        rows[i] = rows[i].join('');
    }
    
    return rows.join('');
};

File: 6.java
// LeetCode Zigzag Conversion
// Reference: https://leetcode.com/problems/zigzag-conversion/solutions/3403/easy-to-understand-java-solution/?envType=study-plan-v2&envId=top-interview-150

class Solution {
    public String convert(String s, int numRows) {
        char[] c = s.toCharArray();

        int len = c.length;
        StringBuffer[] sb = new StringBuffer[numRows];

        for (int i = 0; i < sb.length; i++){
            sb[i] = new StringBuffer();
        }

        int i = 0;
        while (i < len){
            for (int idx = 0; idx < numRows && i < len; idx++){
                // vertically down
                sb[idx].append(c[i++]);
            }
                // obliquely up
            for (int idx = numRows - 2; idx >= 1 && i < len; idx--){

                sb[idx].append(c[i++]);
            }
        }

        for (int idx = 1; idx < sb.length; idx++){
            sb[0].append(sb[idx]);
        }
        return sb[0].toString();
        
    }
}
